src/test/java/dev/ai4j/openai4j/chat/ChatCompletionStreamingTest.java

```diff
@@ -769,45 +767,25 @@ void testCancelStreamingAfterStreamingStarted() throws Exception {
                 .logStreamingResponses()
                 .build();
 
-        AtomicBoolean streamingStarted = new AtomicBoolean(false);
-        AtomicBoolean streamingCancelled = new AtomicBoolean(false);
-        AtomicBoolean cancellationSucceeded = new AtomicBoolean(true);
+        final AtomicBoolean streamingCancelled = new AtomicBoolean(false);
+        final AtomicReference<ResponseHandle> atomicReference = new AtomicReference<>();
+        final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
 
         ResponseHandle responseHandle = client.chatCompletion("Write a poem about AI in 10 words")
                 .onPartialResponse(partialResponse -> {
-                    streamingStarted.set(true);
-                    System.out.println("[[streaming started]]");
-                    if (streamingCancelled.get()) {
-                        cancellationSucceeded.set(false);
-                        System.out.println("[[cancellation failed]]");
+                    if (! streamingCancelled.getAndSet(true)) {
+                        atomicReference.get().cancel();
```

Comment:
IIRC the idea was to test if the cancellation works when done from another thread, so it would be nice to keep the old test. Did you face any issues with it?

Reply:
The **new** version is *also* testing that the cancellation works when done from another thread.

In the new code, a separate thread will run the `onPartialResponse` callback. Some OkHttp3 library *executor* thread is effectively cancelling the request issued by the *main* (test) thread, on the 1st callback.

I added logging of the thread id to validate this.

1. Within the main test:
```
2024-08-07 21:07:07 [main] dev.ai4j.openai4j.chat.ChatCompletionStreamingTest.testCancelStreamingAfterStreamingStarted()
INFO: Test thread Thread[#1,main,5,main]
```

2. Within the callback:
```
2024-08-07 21:07:10 [OkHttp https://api.openai.com/...] dev.ai4j.openai4j.chat.ChatCompletionStreamingTest.lambda$testCancelStreamingAfterStreamingStarted$27()
INFO: Executor thread Thread[#33,OkHttp https://api.openai.com/...,5,main]
```

Let me know if this helps or if I'm still missing something. The reason why I refactored the cancellation handshake is because the test would sometimes fails if run multiple times. Here is the assertion failure:

```
java.lang.AssertionError:
Expecting AtomicBoolean(false) to have value:
  true
but did not.

	at dev.ai4j.openai4j.chat.ChatCompletionStreamingTest.testCancelStreamingAfterStreamingStarted(ChatCompletionStreamingTest.java:810)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
```

My system:

```
$ uname -a
Linux osasuna 6.1.0-23-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.99-1 (2024-07-15) x86_64 GNU/Linux

$ /usr/lib/jvm/jdk-22.0.1/bin/java -version
openjdk version "22.0.1" 2024-04-16
OpenJDK Runtime Environment (build 22.0.1+8-16)
OpenJDK 64-Bit Server VM (build 22.0.1+8-16, mixed mode, sharing)
```

Reply:
If there's a concern about cancelling the request on the first callback *within* the same OkHttp3 thread, I can change and schedule a new thread that will then call the cancel method within a *separate* 3rd thread.

As long as we keep this asynchronous. We should avoid introducing artificial sleeps to wait for the request to start before trying to cancel it. Sometimes this would result into issuing the cancel request too late after the completion event, causing the occasional test failure.

Reply:
Yeah, my concern was calling `cancel()` from the same thread that will then check if request was cancelled. I wanted to test that cancellation flag (which is volatile now but might change/refactored in the future) value propagates properly between threads.

Reply:
Now issuing the cancel in a separate thread from the callback thread.

```
2024-08-08 08:22:33 [main] dev.ai4j.openai4j.chat.ChatCompletionStreamingTest.testCancelStreamingAfterStreamingStarted()
INFO: Test thread Thread[#1,main,5,main]

2024-08-08 08:22:34 [OkHttp https://api.openai.com/...] dev.ai4j.openai4j.chat.ChatCompletionStreamingTest.lambda$testCancelStreamingAfterStreamingStarted$28()
INFO: onPartialResponse thread Thread[#33,OkHttp https://api.openai.com/...,5,main]

2024-08-08 08:22:34 [ForkJoinPool.commonPool-worker-1] dev.ai4j.openai4j.chat.ChatCompletionStreamingTest.lambda$testCancelStreamingAfterStreamingStarted$27()
INFO: cancelling thread Thread[#39,ForkJoinPool.commonPool-worker-1,5,main]
```

Using `volatile` for the cancel flag should be fine.